### ISSUE TITLE
issue: unexpected formState error for Controller and custom input

### ISSUE THREAD

**Issue by Razinar91** [react-hook-form/react-hook-form#8373]
### Version Number

7.31.1

### Codesandbox/Expo snack

https://codesandbox.io/s/brave-lamarr-4u1zyx

### Steps to reproduce

1. npx create-react-app test-hook-form --template typescript
2. npm install react-hook-form
3. Replace the content of App.tsx with the following:

```
import { Controller, useForm } from "react-hook-form";
import InputTest from "./InputTest";

export const App: React.FC = () => {
  const { handleSubmit, formState, control } = useForm({
    mode: "onChange",
  });

  const onChangeHandler = (data: any) => {
    console.log(formState.errors);
  };

  const onSubmit = (data: any) => {
    console.log("submitting...", data);
    console.log("errors...", formState.errors);
  };

  return (
    <div className="App">
      <form onSubmit={handleSubmit(onSubmit)}>
        <Controller
          control={control}
          name="testinput"
          rules={{ required: true, minLength: 5 }}
          render={({ field: { onChange, onBlur, value, ref } }) => (
            <InputTest
              label="testing"
              id="123"
              value={value}
              onChange={(e) => {
                onChange(e.target.value);
                onChangeHandler(e);
              }}
            ></InputTest>
          )}
        />
        <input type="submit" />
      </form>
    </div>
  );
};

export default App;
```

5. Create a new component src/InputTest.tsx

```
import React, { ChangeEvent } from "react";

type inputTestProps = {
  value: string;
  label: string;
  id: string;
  /** Function executed when input value changes*/
  onChange?: (ev: ChangeEvent<HTMLInputElement>) => void;
};

const InputTest: React.FC<inputTestProps> = (props) => {
  const onChangeHandler = (ev: ChangeEvent<HTMLInputElement>) => {
    if (props.onChange) props.onChange(ev);
  };
  return (
    <>
      <label htmlFor={props.id}>{props.label}</label>
      <input id={props.id} onChange={onChangeHandler} />
    </>
  );
};

export default InputTest;
```

7. npm start
8. Open the browser console
9. Write something in the input to make onChange trigger the console.log
10. The first log is { }

### Expected behaviour

The first log should contain an error like this
```
testinput: {
  type: "required"
  message: ""
  ref: {
    name: "testinput"
  }
}
```

From the sandbox created, this is the only problem. Anyways, if you follow the same step and you try the code locally running npm start , the formState.error logged is really inconsistent and i can't understand why

### What browsers are you seeing the problem on?

Chrome, Edge

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by bluebill1049** [comment]
hey there, it's not very clear on the issue that you try to report here. 

> From the sandbox created, this is the only problem. Anyways, if you follow the same step and you try the code locally running npm start , the formState.error logged is really inconsistent and i can't understand why

what's the real issue here?

**Comment by Razinar91** [comment]
Hi bill, sorry it's the first time i open an issue.
The problem i have is that if you follow the step that i pointed out in the
issue on your local machine and you run npm start, the formState.error
value logged by the onChange event is really inconsistent.

On the sandbox created, it seems that the problem is just on the first log,
that is empty (but the field is required). I'm getting mad trying to make
it work

Il Mer 18 Mag 2022, 12:16 Bill ***@***.***> ha scritto:

**Comment by bluebill1049** [comment]
Your `onChange` handler wouldn't have direct access to the `formState` errors, as `errors `will get flushed in the next render. 

1. updating (onChange)
2. react flush down errors 

the errors will flush in the next render.

If you want to have direct validation result then you can consider to use `validate` function

```tsx
validate: () => {
  // do your validation here, depend on return true of false, errors object will get updated
}
```

**Comment by Razinar91** [comment]
So let's say i want to access the error status inside the onChangeHandler or a onBlurHandler.
Are the values always old?

The problem raised because i want to pass down to the custom InputTest component a boolean prop to display a valid or invalid status, accessing formState.errors.testinput. But with the following the valid value is not consisent

```
        <Controller
          control={control}
          name="testinput"
          rules={{ required: true, minLength: 5 }}
          render={({ field: { onChange, onBlur, value, ref } }) => (
            <InputTest
              valid= {formState.errors.testinput}
              label="testing"
              id="123"
              value={value}
              onChange={(e) => {
                onChange(e.target.value);
                onChangeHandler(e);
              }}
            ></InputTest>
          )}
        />
```

**Comment by bluebill1049** [comment]
> The problem raised because i want to pass down to the custom InputTest component a boolean prop to display a valid or invalid status, accessing formState.errors.testinput. But with the following the valid value is not consistent

no, form state will always return the updated errors state after the user's interaction.  The issue you refer original is within `onChange` callback, as it will be behind the render phase.
